apps/apps.c

```diff
@@ -2716,6 +2716,125 @@ BIO *bio_open_owner(const char *filename, int format, int private)
     return NULL;
 }
 
+/*
+ * Open a file for writing, owner-read-only. The file is written in
+ * a 'write-to-temp-file-then-rename' basis.
+ */
+#define RND_SIZE 16
+
+static const BIO_METHOD *BIO_f_tempfile(void);
+
+static void make_temp_filename(const char *orig, char *filename)
+{
+    int i, size = RND_SIZE;
+    char *p;
+    char set[] = "0123456789abcdefghijklmnopqrstuvwxyz"
+                 "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
+
+    strcpy(filename, orig);
+    p = filename + strlen(orig);
+    *p++ = '.';
+
+    srand(time(NULL));
+    while (size--) {
+        i = rand() % (sizeof(set) - 1);
+        *p++ = set[i];
+    }
+
+    *p = '\0';
+
+    return;
+}
+
+BIO *bio_open_owner_tmp(const char *filename, int format, int private)
+{
+    FILE *fp = NULL;
+    BIO *b = NULL;
+    BIO *sink = NULL;
+    BIO *filter = NULL;
+    int fd = -1, bflags, mode, textmode;
+    char *tmp_filename = NULL;
+
+    if (!private || filename == NULL || strcmp(filename, "-") == 0)
+        return bio_open_default(filename, 'w', format);
+
+    /* create tmp file name */
+    tmp_filename = OPENSSL_malloc(strlen(filename) + RND_SIZE + 2);
+    if (tmp_filename == NULL)
+        goto err;
+    make_temp_filename(filename, tmp_filename);
+
+    /* create a tmp file sink BIO */
+    mode = O_WRONLY;
+#ifdef O_CREAT
+    mode |= O_CREAT;
+#endif
+#ifdef O_TRUNC
+    mode |= O_TRUNC;
```

Comment:
I'm not sure `O_TRUNC` is appropriate here.  This is a temporary file, it shouldn't ever exist should it?

I'd include `O_EXCL` and `O_NOFOLLOW`, the first to guarantee the file doesn't exist and the second to avoid symbolic links (even though this is usually implied by `O_CREAT | O_EXCL`).